src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriteStream.cs

```diff
@@ -0,0 +1,274 @@
+
+            if (isAsync)
+            {
+                await _underlyingStream.WriteAsync(_WriteBuffer, 0, WriteBufferOffset, ct).ConfigureAwait(false);
+                await _underlyingStream.FlushAsync(ct).ConfigureAwait(false);
+            }
+            else
+            {
+                _underlyingStream.Write(_WriteBuffer, 0, WriteBufferOffset);
+                _underlyingStream.Flush();
+            }
+
+            // A hard flush means that the TDS message is terminated.
+            // Reset the packet header type so that the next message sender on the stream
+            // will have to set the packet header type.
+            if (flushMode == FlushMode.HardFlush)
+            {
+                // Reset the packet header type
+                PacketHeaderType = TdsStreamPacketType.None;
+            }
+
+            // Reset the offset since we will start filling up the packet again.
+            WriteBufferOffset = TdsEnums.HEADER_LEN;
+        }
+
+        /// <inheritdoc />
+        public override int Read(byte[] buffer, int offset, int count) => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override void SetLength(long value) => throw new NotSupportedException();
+
+        /// <summary>
+        /// Write operation with the byte buffer is not implemented yet.
+        /// </summary>
+        /// <param name="buffer"></param>
+        /// <param name="offset"></param>
+        /// <param name="count"></param>
+        /// <exception cref="NotSupportedException"></exception>
+        public override void Write(byte[] buffer, int offset, int count) 
+            => Write(buffer.AsSpan(offset, count));
+
+        /// <summary>
+        /// An overload for writing a byte to the stream.
+        /// Use this in case the caller is always going to write asynchronously. 
+        /// Otherwise use the WriteByteAsync because it takes the isAsync flag.
+        /// </summary>
+        /// <param name="value">The byte value to be written to the stream.</param>
+        public override void WriteByte(byte value) => WriteByteAsync(value, isAsync: false, ct: CancellationToken.None).AsTask().Wait();
+
+        /// <summary>
+        /// An async mechanism to write a byte to the stream.
+        /// </summary>
+        /// <param name="value">The byte value to be written</param>
+        /// <param name="isAsync">Indicates if the operation is Async</param>
+        /// <param name="ct">Cancellation token to cancel the async operation.</param>
+        /// <returns></returns>
+        public virtual async ValueTask WriteByteAsync(byte value, bool isAsync, CancellationToken ct)
+        {
+            // If we are already at the end of the buffer, flush the buffer, with a softflush.
+            if (_WriteBuffer.Length - WriteBufferOffset == 0)
```

Comment:
One thought here: if we write the byte first then check if we're at the end of the buffer, we get two benefits:
* When buffer fills, it's flushed instead of waiting for next write (negligible impact)
* We can return the ValueTask from `FlushAsync` instead of awaiting it (negligible impact)

Just something to consider

Reply:
I don't follow. We always need to check if the buffer has space in it before writing, else Flush and write. Otherwise we will be writing past the allocated buffer index.

Reply:
What I'm suggesting is flipping the order of check/write. Eg, right now:
* If space in buffer == 0
  * Flush
* Write byte

My suggestion is:
* Write byte
* If space in buffer == 0 (or pointer == buffer.length)
  *  Flush

Though this would also necessitate having the write method also flip its behavior. But, it's mostly negligible impact, afaict. I'd have to run a benchmark to see if there's any impact.

Reply:
How do you write to a buffer which doesn't have space in it? 
If I write a byte if there is no space, then where does it go ?